server/src/server.ts

```diff
@@ -177,6 +183,90 @@ connection.onRequest(ExtensionServerRequest.BundleModules, async ({ workspaceUri
+    return (needFixNodes.length > 0)
+  }
+
+  function buildNewIdAttribute(prefixId: string, id: number): string {
+    const result = prefixId + padLeft(String(id), '0', 5)
+    return result
+  }
+
+  function isIdAttributeExisting(doc: Document, tag: string, prefixId: string, id: number): boolean {
+    const newId = buildNewIdAttribute(prefixId, id)
+    const checkElements = select('//cnxml:' + tag + '[@id = "' + newId + '"]', doc) as Element[]
+    return (checkElements.length > 0)
+  }
+
+  function fixIds(doc: Document, tag: string, prefixId: string): void {
+    const fixNodes = select('//cnxml:' + tag + '[not(@id)]', doc) as Element[]
+    for (const fixNode of fixNodes) {
+      let newId: number = 1
+      while (isIdAttributeExisting(doc, tag, prefixId, newId)) {
+        newId++
+      }
+      const newIdAttribute = buildNewIdAttribute(prefixId, newId)
+      fixNode.setAttribute('id', newIdAttribute)
+    }
+  }
+
+  async function fixModule(moduleName: string): Promise<void> {
+    // 3 steps necessary for element id creation: check, fix, save
+    // == check xml ==
+    // TODO: use cached book-bundle doc data in future for performance increase?
+    const modulePath = path.join(bundle.moduleDirectory(), moduleName, 'index.cnxml')
+    const data = await fs.promises.readFile(modulePath, { encoding: 'utf-8' })
+    const doc = new DOMParser().parseFromString(data)
+    // == fix xml ==
+    if (needFixes(doc)) {
+      fixIds(doc, 'para', 'para')
+      fixIds(doc, 'equation', 'eq')
+      fixIds(doc, 'list', 'list')
+      fixIds(doc, 'item', 'item')
+      fixIds(doc, 'section', 'sect')
+      fixIds(doc, 'problem', 'prob')
+      fixIds(doc, 'solution', 'sol')
+      fixIds(doc, 'exercise', 'exer')
+      fixIds(doc, 'example', 'exam')
+      fixIds(doc, 'figure', 'fig')
+      fixIds(doc, 'definition', 'def')
+      fixIds(doc, 'term', 'term')
+      fixIds(doc, 'meaning', 'mean')
+      fixIds(doc, 'table', 'table')
+      fixIds(doc, 'quote', 'quote')
+      fixIds(doc, 'note', 'note')
+      fixIds(doc, 'footnote', 'foot')
+      fixIds(doc, 'cite', 'cite')
+    }
+    // == save xml ==
+    const out = new XMLSerializer().serializeToString(doc)
+    await fs.promises.writeFile(modulePath, out, { encoding: 'utf-8' })
+  }
+
+  const bundleAndValidator = workspaceBookBundles.get(workspaceUri)
+  // TODO: rework return value on failing?
+  if (bundleAndValidator == null) { return undefined }
+  const bundle = bundleAndValidator[0]
+  const modules = bundle.modules()
+  const orphanModules = Array.from((await bundle.orphanedModules()).inner)
+  const allModules = modules.concat(orphanModules)
+  // TODO: fix modules in parallel. Problem: Could be a memory hog.
+  for (const moduleName of allModules) {
+    await fixModule(moduleName)
```

Comment:
We already discussed this, but leaving as a note for future reference (either to be addressed in this PR or a subsequent issue): I'm curious as to the mileage we'd get in terms of latency decrease if this was async and collected / waited on `Promise.all` instead.